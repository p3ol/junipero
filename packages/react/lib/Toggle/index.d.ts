import { ComponentPropsWithRef, MutableRefObject, ReactNode } from 'react';
import { ForwardedProps } from '../utils';
export declare type ToggleRef = {
    checked: boolean;
    isJunipero: boolean;
    innerRef: MutableRefObject<any>;
    inputRef: MutableRefObject<any>;
};
declare interface ToggleProps extends ComponentPropsWithRef<any> {
    checked?: boolean;
    children?: ReactNode | JSX.Element;
    className?: string;
    disabled?: boolean;
    tabIndex?: number;
    value?: any;
    onChange?(field: {
        value: any;
        checked: boolean;
    }): void;
    ref?: MutableRefObject<ToggleRef | undefined>;
}
declare const Toggle: ForwardedProps<ToggleProps, ToggleRef>;
export default Toggle;
