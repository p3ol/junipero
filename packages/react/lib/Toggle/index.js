import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import PropTypes from 'prop-types';
import { classNames, mockState } from '@junipero/core';

const Toggle = forwardRef(({
  checked = false,
  disabled = false,
  dirty = false,
  children,
  className,
  value,
  onChange,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();

  const [state, dispatch] = useReducer(mockState, {
    checked,
    dirty,
  });

  useEffect(() => {
    dispatch({ checked });
  }, [checked]);

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    isJunipero: true,
  }));

  const onKeyPress = e => {
    if (
      e.key === 'Enter' || e.key === ' '
    ) {
      onChange_(e);
    }
  };

  const onChange_ = () => {
    if (disabled) {
      /* istanbul ignore next: canoot be tested */
      return;
    }

    state.checked = !state.checked;

    dispatch({ checked: state.checked, dirty: true });
    onChange?.({ value, checked: state.checked });
  };

  return (
    <label
      ref={innerRef}
      className={classNames(
        'junipero',
        'toggle-field', {
          disabled,
          checked: state.checked,
        },
        state.dirty ? 'dirty' : 'pristine',
        className
      )}
      tabIndex={1}
      onKeyPress={onKeyPress}
    >
      <input
        { ...rest }
        ref={inputRef}
        type="checkbox"
        value={value}
        disabled={disabled}
        checked={state.checked}
        onChange={onChange_}
        tabIndex={-1}
      />
      <div className="inner">
        <div className="handle" />
      </div>
      <div className="label">
        {children}
      </div>
    </label>
  );
});

Toggle.displayName = 'Toggle';
Toggle.propTypes = {
  value: PropTypes.any,
  checked: PropTypes.bool,
  valid: PropTypes.bool,
  dirty: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  onValidate: PropTypes.func,
  onChange: PropTypes.func,
};
export default Toggle;
