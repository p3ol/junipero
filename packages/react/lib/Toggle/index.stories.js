import { action } from '@storybook/addon-actions';
import { useState } from 'react';

import ToggleField from '.';

export default { title: 'react/ToggleField' };
export const basic = () => (
  <ToggleField>Check that</ToggleField>
);

export const alreadyChecked = () => (
  <ToggleField checked={true}>Uncheck this</ToggleField>
);

export const withOnChange = () => (
  <ToggleField
    onChange={action('change')}
    value="agreement"
  >
    Check this
  </ToggleField>
);

export const disabled = () => (
  <ToggleField disabled={true}>cannot check this</ToggleField>
);

export const disabledChecked = () => (
  <ToggleField disabled={true} checked={true}>cannot uncheck this</ToggleField>
);

export const withValueChanging = () => {
  const [value, setValue] = useState(false);

  const changeValue = () => {
    setValue(!value);
  };

  return (
    <>
      <button onClick={changeValue}>change value</button>
      <ToggleField
        className="boxed"
        checked={value}
        onChange={e => setValue(e.value)}
      >
        Display more
      </ToggleField>
    </>
  );
};
