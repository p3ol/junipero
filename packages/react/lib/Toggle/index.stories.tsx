import { useState } from 'react';
import { action } from '@storybook/addon-actions';

import Toggle from '.';

export default { title: 'react/Toggle' };
export const basic = () => (
  <Toggle>Check that</Toggle>
);

export const alreadyChecked = () => (
  <Toggle checked={true}>Uncheck this</Toggle>
);

export const withOnChange = () => (
  <Toggle
    onChange={action('change')}
    value="agreement"
  >
    Check this
  </Toggle>
);

export const disabled = () => (
  <Toggle disabled={true}>cannot check this</Toggle>
);

export const disabledChecked = () => (
  <Toggle disabled={true} checked={true}>cannot uncheck this</Toggle>
);

export const withValueChanging = () => {
  const [value, setValue] = useState(false);

  const changeValue = () => {
    setValue(!value);
  };

  return (
    <>
      <button onClick={changeValue}>change value</button>
      <Toggle
        className="boxed"
        checked={value}
        onChange={e => setValue(e.value)}
      >
        Display more
      </Toggle>
    </>
  );
};
