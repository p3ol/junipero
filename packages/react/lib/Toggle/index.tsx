import {
  type KeyboardEvent,
  type RefObject,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';

export declare type ToggleValue = any;

export declare interface ToggleRef extends JuniperoRef {
  checked: boolean;
  innerRef: RefObject<HTMLLabelElement>;
  inputRef: RefObject<HTMLInputElement>;
}

export declare interface ToggleProps extends Omit<
  SpecialComponentPropsWithRef<'input', ToggleRef>,
  'onChange'
> {
  checked?: boolean;
  disabled?: boolean;
  tabIndex?: number;
  value?: ToggleValue;
  onChange?(field: FieldContent<ToggleValue>): void;
}

export declare interface ToggleState {
  checked: boolean;
}

const Toggle = ({
  ref,
  checked = false,
  disabled = false,
  tabIndex = 1,
  children,
  className,
  value,
  onChange,
  ...rest
}: ToggleProps) => {
  const innerRef = useRef<HTMLLabelElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const [state, dispatch] = useReducer(mockState<ToggleState>, {
    checked,
  });

  useEffect(() => {
    dispatch({ checked });
  }, [checked]);

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    checked: state.checked,
    isJunipero: true,
  }));

  const onKeyDown = (e: KeyboardEvent) => {
    if (e.key === 'Enter' || e.key === ' ') {
      onChange_();
    }
  };

  const onChange_ = () => {
    if (disabled) {
      /* istanbul ignore next: canoot be tested */
      return;
    }

    state.checked = !state.checked;

    dispatch({ checked: state.checked });
    onChange?.({ value, checked: state.checked });
  };

  return (
    <label
      ref={innerRef}
      className={classNames(
        'junipero',
        'toggle-field',
        {
          disabled,
          checked: state.checked,
        },
        className
      )}
      tabIndex={tabIndex}
      onKeyDown={onKeyDown}
    >
      <input
        { ...rest }
        ref={inputRef}
        type="checkbox"
        value={value}
        disabled={disabled}
        checked={state.checked}
        onChange={onChange_}
        tabIndex={-1}
      />
      <div className="inner">
        <div className="handle" />
      </div>
      { children && (
        <div className="label">
          { children }
        </div>
      ) }
    </label>
  );
};

Toggle.displayName = 'Toggle';

export default Toggle;
