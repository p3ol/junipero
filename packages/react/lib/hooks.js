import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
