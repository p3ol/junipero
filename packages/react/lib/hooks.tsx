import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
