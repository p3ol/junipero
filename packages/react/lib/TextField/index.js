import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
