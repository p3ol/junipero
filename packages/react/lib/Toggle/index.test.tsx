import { render, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Toggle from '.';

describe('<Toggle />', () => {
  it('should render', () => {
    const { container, unmount } = render(<Toggle />);

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should render with children', () => {
    const { container, unmount } = render(
      <Toggle>
        label
      </Toggle>
    );

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should set disabled class if toggle is disabled', () => {
    const { container, unmount } = render(
      <Toggle disabled={true}>
        label
      </Toggle>
    );

    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should handle change event', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const { container, unmount } = render(
      <Toggle
        onChange={onChangeMock}
        value="my_value"
      >
        label
      </Toggle>
    );
    expect(container).toMatchSnapshot('unChecked');
    const input = container.querySelector('input');
    user.click(input);
    await waitFor(() => expect(onChangeMock).toHaveBeenCalledWith(
      { value: 'my_value', checked: true }
    ));

    expect(container).toMatchSnapshot('checked');
    unmount();
  });

  it('should handle change event with space and enter ' +
  'when focused', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const { container, unmount } = render(
      <Toggle
        onChange={onChangeMock}
        value="my_value"
      >
        label
      </Toggle>
    );
    expect(container).toMatchSnapshot('unChecked');
    await act(async () => {
      return Promise.resolve(container.querySelector('label').focus());
    });
    user.keyboard('{ }');
    await waitFor(() => expect(onChangeMock).toHaveBeenCalledWith(
      { value: 'my_value', checked: true }
    ));

    expect(container).toMatchSnapshot('checked');
    user.keyboard('{Enter}');
    await waitFor(() => expect(onChangeMock).toHaveBeenCalledWith(
      { value: 'my_value', checked: false }
    ));

    expect(container).toMatchSnapshot('unchecked');

    unmount();
  });

  it('should not handle change if toggle is disabled', () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const { container, unmount } = render(
      <Toggle
        onChange={onChangeMock}
        disabled={true}
        value="my_value"
      >
        label
      </Toggle>
    );
    expect(container).toMatchSnapshot('unChecked');
    const input = container.querySelector('input');
    user.click(input);
    expect(onChangeMock).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot('unchecked');
    unmount();
  });
});
