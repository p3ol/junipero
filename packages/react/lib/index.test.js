import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
