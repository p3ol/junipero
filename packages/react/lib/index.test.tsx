import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
