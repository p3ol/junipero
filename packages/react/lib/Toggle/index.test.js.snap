// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should handle change event with space and enter when focused: checked and dirty 1`] = `
<div>
  <label
    class="junipero toggle-field checked dirty"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should handle change event with space and enter when focused: unChecked and pristine 1`] = `
<div>
  <label
    class="junipero toggle-field pristine"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should handle change event with space and enter when focused: unchecked and dirty 1`] = `
<div>
  <label
    class="junipero toggle-field dirty"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should handle change event: checked and dirty 1`] = `
<div>
  <label
    class="junipero toggle-field checked dirty"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should handle change event: unChecked and pristine 1`] = `
<div>
  <label
    class="junipero toggle-field pristine"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should not handle change if toggle is disabled: unChecked and pristine 1`] = `
<div>
  <label
    class="junipero toggle-field disabled pristine"
    tabindex="1"
  >
    <input
      disabled=""
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should not handle change if toggle is disabled: unchecked and pristine 1`] = `
<div>
  <label
    class="junipero toggle-field disabled pristine"
    tabindex="1"
  >
    <input
      disabled=""
      tabindex="-1"
      type="checkbox"
      value="my_value"
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should render 1`] = `
<div>
  <label
    class="junipero toggle-field pristine"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value=""
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    />
  </label>
</div>
`;

exports[`<Toggle /> should render with children 1`] = `
<div>
  <label
    class="junipero toggle-field pristine"
    tabindex="1"
  >
    <input
      tabindex="-1"
      type="checkbox"
      value=""
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;

exports[`<Toggle /> should set disabled class if toggle is disabled 1`] = `
<div>
  <label
    class="junipero toggle-field disabled pristine"
    tabindex="1"
  >
    <input
      disabled=""
      tabindex="-1"
      type="checkbox"
      value=""
    />
    <div
      class="inner"
    >
      <div
        class="handle"
      />
    </div>
    <div
      class="label"
    >
      label
    </div>
  </label>
</div>
`;
